File: SUSTech/CS203/2019Fall/Lab3/C.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
const int maxn = 3e3 + 5;

int n, m;
ll ans = 0;
int a[maxn];

int main() {
#ifdef DEBUG
    freopen("test.in", "r", stdin);
    freopen("test.out", "w", stdout);
#endif
    ios::sync_with_stdio(false);
    cin.tie(0);

    cin >> n >> m;
    for (int i = 1; i <= n; ++i) cin >> a[i];
    sort(a + 1, a + n + 1);
    for (int i = 1; i <= n; ++i) {
        for (int j = i + 1; j <= n; ++j)
            ans += upper_bound(a + j + 1, a + n + 1, m - a[i] - a[j]) -
                   lower_bound(a + j + 1, a + n + 1, m - a[i] - a[j]);
    }
    cout << ans << '\n';

    return 0;
}

File: SUSTech/CS203/2019Fall/Lab3/C.java
import java.util.*;

public class C {

    static int n, m;
    static long ans = 0;

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        n = in.nextInt();
        m = in.nextInt();
        int[] a = new int[n];
        for (int i = 0; i < n; ++i)
            a[i] = in.nextInt();
        in.close();
        Arrays.sort(a);
        for (int i = 0; i < n; ++i)
            for (int j = i + 1; j < n; ++j)
                ans += upperBound(a, j + 1, n - 1, m - a[i] - a[j]) - lowerBound(a, j + 1, n - 1, m - a[i] - a[j]);
        System.out.println(ans);
    }

    public static int lowerBound(int[] nums, int l, int r, int target) {
        int ans = n;
        while (l <= r) {
            int m = (l + r) / 2;
            if (nums[m] >= target) {
                ans = m;
                r = m - 1;
            } else
                l = m + 1;
        }
        return ans;
    }

    public static int upperBound(int[] nums, int l, int r, int target) {
        int ans = n;
        while (l <= r) {
            int m = (l + r) / 2;
            if (nums[m] <= target)
                l = m + 1;
            else {
                ans = m;
                r = m - 1;
            }
        }
        return ans;
    }
}